File: src/main/java/edu/uchicago/gerber/_08final/mvc/controller/GameOpsQueue.java
package edu.uchicago.gerber._08final.mvc.controller;

import edu.uchicago.gerber._08final.mvc.model.Movable;

import java.util.*;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;


/**
 * Effectively a Queue that enqueues and dequeues Game Operations (add/remove)
 */
public class GameOpsQueue extends LinkedList<GameOp> {

    public void enqueue(Movable mov, GameOp.Action action) {
        addLast(new GameOp(mov, action));
    }

    public GameOp dequeue() {
        return removeFirst();
    }
}

File: src/main/java/edu/uchicago/gerber/_08final/mvc/model/GameOpsQueue.kt
package edu.uchicago.gerber._08final.mvc.model

import java.util.*
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

/**
 * Created by ag on 6/17/2015.
 */
class GameOpsQueue : LinkedList<GameOp>() {
    //this data structure is in contention by the "Event Dispatch" thread aka main-swing-thread, and the animation
    // thread. We must restrict access to it by one thread at a time by using a Lock.
    private val lock: Lock

    init {
        lock = ReentrantLock()
    }

    fun enqueue(mov: Movable, action: GameOp.Action) {
        try {
            lock.lock()
            addLast(GameOp(mov, action))
        } finally {
            lock.unlock()
        }
    }

    fun dequeue(): GameOp? {
        return try {
            lock.lock()
            removeFirst()
        } finally {
            lock.unlock()
        }
    }
}
